File: sdk/python/tonebridge/services/analyze.py
"""
Analyze Service
Handles text analysis operations
"""

from typing import Optional, Dict, Any, List, Union
from ..types import (
    AnalysisType,
    Priority,
    AnalyzeRequest,
    AnalyzeResponse,
)
from ..constants import API_ENDPOINTS, MAX_TEXT_LENGTH
from ..exceptions import ValidationError


class AnalyzeService:
    """Service for text analysis operations"""
    
    def __init__(self, client):
        """
        Initialize analyze service
        
        Args:
            client: ToneBridgeClient instance
        """
        self.client = client
    
    def analyze(
        self,
        text: str,
        analysis_types: Optional[List[Union[AnalysisType, str]]] = None,
        include_suggestions: bool = False,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Analyze text
        
        Args:
            text: Text to analyze
            analysis_types: Types of analysis to perform
            include_suggestions: Include improvement suggestions
            metadata: Additional metadata
            
        Returns:
            Analysis results
            
        Example:
            >>> result = client.analyze.analyze(
            ...     "This is urgent!",
            ...     analysis_types=[AnalysisType.TONE, AnalysisType.PRIORITY],
            ...     include_suggestions=True
            ... )
        """
        self._validate_analyze_request(text)
        
        request_data = {
            "text": text,
            "include_suggestions": include_suggestions,
        }
        
        if analysis_types:
            request_data["analysis_types"] = [
                t.value if isinstance(t, AnalysisType) else t
                for t in analysis_types
            ]
        
        if metadata:
            request_data["metadata"] = metadata
        
        return self.client.request("POST", API_ENDPOINTS["ANALYZE"], json=request_data)
    
    def analyze_tone(self, text: str) -> str:
        """
        Analyze text tone
        
        Args:
            text: Text to analyze
            
        Returns:
            Tone of the text
        """
        result = self.analyze(text, analysis_types=[AnalysisType.TONE])
        return result.get("data", {}).get("tone", "unknown")
    
    def analyze_clarity(self, text: str) -> float:
        """
        Analyze text clarity
        
        Args:
            text: Text to analyze
            
        Returns:
            Clarity score (0-100)
        """
        result = self.analyze(text, analysis_types=[AnalysisType.CLARITY])
        return result.get("data", {}).get("clarity_score", 0.0)
    
    def analyze_priority(self, text: str) -> Dict[str, str]:
        """
        Analyze text priority
        
        Args:
            text: Text to analyze
            
        Returns:
            Priority and quadrant information
        """
        result = self.analyze(text, analysis_types=[AnalysisType.PRIORITY])
        data = result.get("data", {})
        return {
            "priority": data.get("priority", "medium"),
            "quadrant": data.get("priority_quadrant"),
        }
    
    def analyze_sentiment(self, text: str) -> Dict[str, float]:
        """
        Analyze text sentiment
        
        Args:
            text: Text to analyze
            
        Returns:
            Sentiment polarity and subjectivity
        """
        result = self.analyze(text, analysis_types=[AnalysisType.SENTIMENT])
        return result.get("data", {}).get("sentiment", {})
    
    def comprehensive_analysis(
        self,
        text: str,
        include_suggestions: bool = True,
    ) -> Dict[str, Any]:
        """
        Perform comprehensive analysis with all types
        
        Args:
            text: Text to analyze
            include_suggestions: Include improvement suggestions
            
        Returns:
            Comprehensive analysis results
        """
        return self.analyze(
            text,
            analysis_types=[
                AnalysisType.TONE,
                AnalysisType.CLARITY,
                AnalysisType.PRIORITY,
                AnalysisType.SENTIMENT,
                AnalysisType.STRUCTURE,
                AnalysisType.COMPLETENESS,
            ],
            include_suggestions=include_suggestions,
        )
    
    def score_priority(
        self,
        text: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Score priority using Eisenhower Matrix
        
        Args:
            text: Text to score
            context: Additional context
            
        Returns:
            Priority score and quadrant information
        """
        request_data = {"text": text}
        if context:
            request_data["context"] = context
        
        return self.client.request(
            "POST",
            API_ENDPOINTS["SCORE_PRIORITY"],
            json=request_data
        )
    
    def batch_score_priorities(
        self,
        messages: List[Dict[str, Any]],
    ) -> List[Dict[str, Any]]:
        """
        Batch score priorities for multiple messages
        
        Args:
            messages: List of messages to score
            
        Returns:
            List of priority scores
        """
        return self.client.request(
            "POST",
            API_ENDPOINTS["BATCH_SCORE_PRIORITIES"],
            json={"messages": messages}
        )
    
    def check_needs_transformation(self, text: str) -> Dict[str, Any]:
        """
        Check if text needs transformation
        
        Args:
            text: Text to check
            
        Returns:
            Recommendation for transformation
        """
        analysis = self.comprehensive_analysis(text, include_suggestions=True)
        data = analysis.get("data", {})
        
        needs_transformation = (
            data.get("clarity_score", 100) < 70 or
            data.get("tone") in ["harsh", "technical"] or
            data.get("sentiment", {}).get("polarity", 0) < -0.3
        )
        
        reasons = []
        if data.get("clarity_score", 100) < 70:
            reasons.append("Low clarity score")
        if data.get("tone") == "harsh":
            reasons.append("Harsh tone detected")
        if data.get("tone") == "technical":
            reasons.append("Technical language detected")
        if data.get("sentiment", {}).get("polarity", 0) < -0.3:
            reasons.append("Negative sentiment")
        
        recommended_type = "soften"
        if data.get("clarity_score", 100) < 60:
            recommended_type = "clarify"
        elif data.get("tone") == "technical":
            recommended_type = "terminology"
        
        return {
            "needs_transformation": needs_transformation,
            "recommended_type": recommended_type if needs_transformation else None,
            "confidence": 0.8 if needs_transformation else 0.2,
            "reasons": reasons,
        }
    
    def get_suggestions(self, text: str) -> List[str]:
        """
        Get improvement suggestions for text
        
        Args:
            text: Text to get suggestions for
            
        Returns:
            List of suggestions
        """
        result = self.analyze(text, include_suggestions=True)
        return result.get("data", {}).get("suggestions", [])
    
    def _validate_analyze_request(self, text: str) -> None:
        """Validate analyze request parameters"""
        if not text:
            raise ValidationError("Text is required")
        
        if len(text) > MAX_TEXT_LENGTH:
            raise ValidationError(f"Text exceeds maximum length of {MAX_TEXT_LENGTH} characters")

File: sdk/javascript/src/services/analyze.ts
/**
 * Analyze Service
 * Handles text analysis operations
 */

import { AxiosInstance } from 'axios';
import {
  AnalyzeRequest,
  AnalyzeResponse,
  AnalysisType,
  Priority
} from '../types';
import { API_ENDPOINTS, MAX_TEXT_LENGTH } from '../constants';
import { ValidationError } from '../errors';

export class AnalyzeService {
  constructor(private axios: AxiosInstance) {}

  /**
   * Analyze text
   * @param request Analyze request
   * @returns Analysis results
   * @example
   * ```typescript
   * const analysis = await client.analyze.analyze({
   *   text: "This is an urgent request",
   *   analysis_types: ['tone', 'priority'],
   *   include_suggestions: true
   * });
   * ```
   */
  async analyze(request: AnalyzeRequest): Promise<AnalyzeResponse> {
    this.validateAnalyzeRequest(request);

    const response = await this.axios.post<AnalyzeResponse>(
      API_ENDPOINTS.ANALYZE,
      request
    );

    return response.data;
  }

  /**
   * Quick analyze for tone
   * @param text Text to analyze
   */
  async analyzeTone(text: string): Promise<string> {
    const result = await this.analyze({
      text,
      analysis_types: ['tone']
    });

    return result.data.tone;
  }

  /**
   * Quick analyze for clarity
   * @param text Text to analyze
   */
  async analyzeClarity(text: string): Promise<number> {
    const result = await this.analyze({
      text,
      analysis_types: ['clarity']
    });

    return result.data.clarity_score;
  }

  /**
   * Quick analyze for priority
   * @param text Text to analyze
   */
  async analyzePriority(text: string): Promise<{
    priority: Priority;
    quadrant?: string;
  }> {
    const result = await this.analyze({
      text,
      analysis_types: ['priority']
    });

    return {
      priority: result.data.priority,
      quadrant: result.data.priority_quadrant
    };
  }

  /**
   * Quick analyze for sentiment
   * @param text Text to analyze
   */
  async analyzeSentiment(text: string): Promise<{
    polarity: number;
    subjectivity: number;
  }> {
    const result = await this.analyze({
      text,
      analysis_types: ['sentiment']
    });

    return result.data.sentiment;
  }

  /**
   * Comprehensive analysis with all types
   * @param text Text to analyze
   * @param includeSuggestions Include improvement suggestions
   */
  async comprehensiveAnalysis(
    text: string,
    includeSuggestions: boolean = true
  ): Promise<AnalyzeResponse> {
    return this.analyze({
      text,
      analysis_types: [
        'tone',
        'clarity',
        'priority',
        'sentiment',
        'structure',
        'completeness'
      ],
      include_suggestions: includeSuggestions
    });
  }

  /**
   * Score priority using Eisenhower Matrix
   * @param text Text to score
   * @param context Additional context
   */
  async scorePriority(
    text: string,
    context?: Record<string, any>
  ): Promise<{
    score: number;
    quadrant: string;
    urgency: number;
    importance: number;
    reasoning: string;
  }> {
    const response = await this.axios.post(
      API_ENDPOINTS.SCORE_PRIORITY,
      {
        text,
        context
      }
    );

    return response.data;
  }

  /**
   * Batch score priorities for multiple messages
   * @param messages Array of messages to score
   */
  async batchScorePriorities(
    messages: Array<{ id: string; text: string; context?: any }>
  ): Promise<Array<{
    id: string;
    score: number;
    quadrant: string;
    ranking: number;
  }>> {
    const response = await this.axios.post(
      API_ENDPOINTS.BATCH_SCORE_PRIORITIES,
      {
        messages
      }
    );

    return response.data;
  }

  /**
   * Check if message needs transformation
   * @param text Text to check
   * @returns Recommendation for transformation
   */
  async checkNeedsTransformation(text: string): Promise<{
    needs_transformation: boolean;
    recommended_type?: string;
    confidence: number;
    reasons: string[];
  }> {
    const analysis = await this.comprehensiveAnalysis(text, true);
    
    const needsTransformation = 
      analysis.data.clarity_score < 70 ||
      analysis.data.tone === 'harsh' ||
      analysis.data.tone === 'technical' ||
      (analysis.data.sentiment?.polarity || 0) < -0.3;

    const reasons = [];
    if (analysis.data.clarity_score < 70) {
      reasons.push('Low clarity score');
    }
    if (analysis.data.tone === 'harsh') {
      reasons.push('Harsh tone detected');
    }
    if (analysis.data.tone === 'technical') {
      reasons.push('Technical language detected');
    }
    if ((analysis.data.sentiment?.polarity || 0) < -0.3) {
      reasons.push('Negative sentiment');
    }

    let recommendedType = 'soften';
    if (analysis.data.clarity_score < 60) {
      recommendedType = 'clarify';
    } else if (analysis.data.tone === 'technical') {
      recommendedType = 'terminology';
    }

    return {
      needs_transformation: needsTransformation,
      recommended_type: needsTransformation ? recommendedType : undefined,
      confidence: needsTransformation ? 0.8 : 0.2,
      reasons
    };
  }

  /**
   * Get improvement suggestions
   * @param text Text to get suggestions for
   */
  async getSuggestions(text: string): Promise<string[]> {
    const result = await this.analyze({
      text,
      include_suggestions: true
    });

    return result.data.suggestions || [];
  }

  /**
   * Validate analyze request
   */
  private validateAnalyzeRequest(request: AnalyzeRequest): void {
    if (!request.text) {
      throw new ValidationError('Text is required');
    }

    if (request.text.length > MAX_TEXT_LENGTH) {
      throw new ValidationError(
        `Text exceeds maximum length of ${MAX_TEXT_LENGTH} characters`
      );
    }
  }
}